[LOLIPOP-INTELLIGENCE] iP

LOLIPOP-INTELLIGENCE opened the pull request:
# 🍎 Appleaster: Task Manager Update

## Changes:
1. Enhanced task creation
2. Improved deadline tracking


Practise your Java:
```java
    public static void main(String[] args) {
        new Appleaster("data/tasks.txt").run();
    }
```

- [x] Complete tests
- [ ] Update docs

> Simplify your life with Appleaster!

**Key feature**: *Natural language* processing for task input.

~~Minor updates~~ Major improvements!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

manas-sarvam committed:
Adds initial welcome message and changes bot name

manas-sarvam committed:
adds echo messages

manas-sarvam committed:
completes level 2 requirements

manas-sarvam committed:
adds level-3 requirements

manas-sarvam committed:
adds level 4 requirements

manas-sarvam committed:
adds initial test

manas-sarvam committed:
adds primitive exception handling

manas-sarvam committed:
adds delete functionality

manas-sarvam committed:
adds enums for tasks and commands

manas-sarvam committed:
Implements saving of list items to local storage

manas-sarvam committed:
merge branch-Level-7 to master

manas-sarvam committed:
Add datetime functionality

manas-sarvam committed:
Merge branch 'branch-Level-8'

manas-sarvam committed:
Add Storage, Ui, Parser and TaskList

manas-sarvam committed:
Refactor code into packages

manas-sarvam committed:
Merge remote-tracking branch 'origin/add-gradle-support'

manas-sarvam committed:
Add gradle support to build and run Appleaster

manas-sarvam committed:
Add JUnit test for Parser and TaskList

manas-sarvam committed:
Create Jar file

manas-sarvam committed:
Add JavaDoc comments for 3 Parser, Appleaster and TaskList

manas-sarvam committed:
Refactor code to coding standard

manas-sarvam committed:
Add find command

manas-sarvam committed:
Refactor code to meet coding standard

manas-sarvam committed:
Add find feature

Edsel-Tan commented on src/main/java/com/appleaster/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+        String commandWord = parts[0].toLowerCase();
+        String arguments = parts.length > 1 ? parts[1] : "";
+
+    switch (commandWord) {
```
Wrong indentation level?

Edsel-Tan commented on src/main/java/com/appleaster/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+    switch (commandWord) {
+      case "list":
+        return new Command(CommandType.LIST);
+      case "mark":
```
Is the fallthrough intentional? Should leave a comment

Edsel-Tan commented on src/main/java/com/appleaster/main/Appleaster.java:
```diff
@@ -0,0 +1,107 @@
+    private boolean executeCommand(Command c) {
+        try {
+            switch (c.getType()) {
+                case LIST:
```
wrong indentation level

blanklogic replied:
The use of switch case code blocks do not require indentations according to SE-EDU Java Coding Standard, refer to the example below.

```java
switch (c.getType()) {
case LIST:
    task.listTask();
    break;
case MARK:
case UNMARK:
    /* Code omitted ... */
```

Edsel-Tan reviewed (commented):
Code is super nice and clean! 👍

blanklogic commented on data/tasks.txt:
```diff
@@ -0,0 +1,2 @@
+T | 1 | call mom
```
As data/tasks.txt is going to be constantly changing, might be good to not track it. Consider adding it to gitignore as well as removing this cached version to prevent any further tracking by git.

blanklogic commented on src/main/java/com/appleaster/command/CommandType.java:
```diff
@@ -0,0 +1,14 @@
+package com.appleaster.command;
+public enum CommandType {
```
Nice utilisation of Enums to simplify the finite Command Types.

blanklogic commented on src/main/java/com/appleaster/command/Command.java:
```diff
@@ -0,0 +1,56 @@
+
+public class Command {
+    private final CommandType type;
+    private Task task;
```
The following fields could be effectively final as well, I noticed the rest of your code in this class do not further modify them after being initialised by the use of overloaded constructors

blanklogic commented on src/main/java/com/appleaster/command/Command.java:
```diff
@@ -0,0 +1,56 @@
+        return type;
+    }
+
+    public Task getTask() {
```
# Potential NullPointerException
- Since your overloaded constructors do not initialise the values of the other fields, only specific ones.
- Calling upon getTask() on a Command object constructed with public Command(CommandType type) might throw a NullPointerException.
- Consider including default values for the other fields if not in use to prevent this.

blanklogic commented on src/main/java/com/appleaster/exception/AppleasterException.java:
```diff
@@ -0,0 +1,6 @@
+package com.appleaster.exception;
+public class AppleasterException extends Exception {
```
Perhaps extending from RuntimeException might be more appropriate as AppleasterException is thrown during run time as opposed to during compile time.

blanklogic commented on src/main/java/com/appleaster/main/Appleaster.java:
```diff
@@ -0,0 +1,107 @@
+
+public class Appleaster {
+  private final Storage storage;
+  private TaskList tasks;
```
Can consider being made final as TaskList reference is not modified after being initialised by constructor

blanklogic commented on src/main/java/com/appleaster/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+ * Parses user input into Command objects for the Appleaster application.
+ */
+public class Parser {
+  private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ISO_LOCAL_DATE;
```
Wrong indentation level. Should be an indentation of 4 spaces.

blanklogic commented on src/main/java/com/appleaster/storage/Storage.java:
```diff
@@ -0,0 +1,102 @@
+        Files.createDirectories(filePath.getParent());
+    }
+
+    private String formatTask(Task task) {
```
Might be good to leave the String representation formatting to the classes themselves such as Todo.java, Deadline.java, Event.java. Furthermore, notice how there's a lot of overlaps in the code that have been repeated. Perhaps adding the common formatting to the abstract class Task.java and Overriding the String representation in it's children could make it easier to locate and make changes in the future.

tldr; OOP concept of abstraction barrier. Ask the task to return the String instead of asking information from the task to create your own String.

blanklogic commented on src/main/java/com/appleaster/task/Task.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Override
+    public abstract String toString();
```
As mentioned in another comment, might be good to initialise your toString() in Task.java to contain the key information that is utilised by all the following children Todo.java, Event.java, Deadline.java and Overriding them as necessary to add additional information to be more specific to the child.

blanklogic commented on src/main/java/com/appleaster/task/Task.java:
```diff
@@ -0,0 +1,31 @@
+package com.appleaster.task;
+public abstract class Task {
+    protected String description;
+    protected TaskStatus status;
```
A boolean of isCompleted should suffice instead of polluting the heap with more objects than needed.

blanklogic commented on src/main/java/com/appleaster/task/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Override
+    public String toString() {
```
Utilise the OOP concept of inheritance by abstracting out the common information such as description to be returned by the superclass/parent Task.java instead before Overriding and adding more specific information to reduce the need to modify multiple files for a change.

blanklogic commented on src/main/java/com/appleaster/task/Event.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public String toString() {
```
Utilise the OOP concept of inheritance by abstracting out the common information such as description to be returned by the superclass/parent Task.java instead before Overriding and adding more specific information to reduce the need to modify multiple files for a change.

blanklogic commented on src/main/java/com/appleaster/task/TaskList.java:
```diff
@@ -0,0 +1,169 @@
+package com.appleaster.task;
+
+import com.appleaster.exception.AppleasterException;
```
Might have been a mistake of double importing but the import at Line 9 is appropriate. Line 3 should trigger a checkstyle error.

blanklogic commented on src/main/java/com/appleaster/task/TaskList.java:
```diff
@@ -0,0 +1,169 @@
+     *
+     * @param task The task to be added.
+     */
+  public void addTask(Task task) {
```
Wrong indentation level for all the code blocks after this line.

blanklogic commented on src/main/java/com/appleaster/task/TaskStatus.java:
```diff
@@ -0,0 +1,15 @@
+package com.appleaster.task;
+public enum TaskStatus {
```
Might be redundant to abstract TaskStatus out as an Enum when a isCompleted() and String formatting of X in Task.java should suffice.

blanklogic reviewed (commented):
# Overview
There were quite a few violations of Java Coding Standards in the Code. But with the appropriate fixes suggested should be easy to rectify. Consider drawing out a UML Diagram to better make sense of the code structure for OOP so as to optimise the amount of Objects being created as well as to incorporate Software Engineering Practices such as Do not Repeat Yourself (DRY) and SOLID.

# Conclusion
Overall, after some fixes LGTM.

manas-sarvam committed:
Add code

manas-sarvam committed:
Add assertions to Parser class for improved error detection

This commit introduces assert statements to the Parser class to catch potential programmer errors and document important assumptions about the code's behavior. These changes enhance the robustness of the parsing logic during development and testing phases.

Changes made:
- Added null checks for input parameters in all parsing methods
- Included structural validity checks for command inputs
- Ensured proper command word for mark/unmark operations
- Verified the presence of required parts in deadline and event commands

Note: Assertions are intended for development and testing environments only. They should be enabled using the -ea flag when running the Java VM for testing.

Testing:
- Run the application with assertions enabled
- Test with both valid and invalid inputs
- Verify that AssertionErrors are thrown for violated assumptions
- Ensure that regular error handling (AppleasterExceptions) still functions for invalid user inputs

Remember to keep assertions disabled in production environments.

manas-sarvam committed:
Performs code quality check and removes unnecessary files

manas-sarvam committed:
Add feature to view schedule for a particular day

manas-sarvam committed:
Add README Guide

manas-sarvam committed:
Add README Guide

manas-sarvam committed:
Add test for Parser class

manas-sarvam committed:
Delete data folder

manas-sarvam committed:
Remove unused data files and test folders

manas-sarvam committed:
Modify README.md

manas-sarvam committed:
Refactor image file path

manas-sarvam committed:
A-Release
